[SPARK-22648] [K8S] Spark on Kubernetes - Documentation

foxish opened the pull request:
What changes were proposed in this pull request?

This PR contains documentation on the usage of Kubernetes scheduler in Spark 2.3, and a shell script to make it easier to build docker images required to use the integration. The changes detailed here are covered by https://github.com/apache/spark/pull/19717 and https://github.com/apache/spark/pull/19468 which have merged already.

How was this patch tested?
The script has been in use for releases on our fork. Rest is documentation. 

cc @rxin @mateiz (shepherd)
k8s-big-data SIG members & contributors: @foxish @ash211 @mccheah @liyinan926 @erikerlandson @ssuchter @varunkatta @kimoonkim @tnachen @ifilonenko 
reviewers: @vanzin @felixcheung @jiangxb1987 @mridulm

TODO:
- [x] Add dockerfiles directory to built distribution. (https://github.com/apache/spark/pull/20007)
- [x] Change references to docker to instead say "container" (https://github.com/apache/spark/pull/19995)
- [x] Update configuration table.
- [x] Modify spark.kubernetes.allocation.batch.delay to take time instead of int (#20032)

foxish committed:
Adding running on k8s.md

AmplabJenkins commented:
Can one of the admins verify this patch?

holdenk commented:
Jenkins OK to test

holdenk commented on sbin/build-push-docker-images.sh:
```diff
@@ -0,0 +1,67 @@
+  done
+}
+
+function usage {
```
So likely not for this PR, but I'm wondering for the future what you think about the idea of extending this to package up a usable Python env so we can solve the dependency management issue as well?

Really excited to see the progress :)

foxish replied:
We do already do that [in our fork](https://github.com/apache-spark-on-k8s/spark/blob/branch-2.2-kubernetes/sbin/build-push-docker-images.sh#L24) - and when we get PySpark and R submitted (in Spark 2.4 hopefully), we will extend this script.

jiangxb1987 commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,496 @@
+### Namespaces
+
+Kubernetes has the concept of [namespaces](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/).
+Namespaces are a way to divide cluster resources between multiple users (via resource quota). Spark on Kubernetes can
```
"a way" -> "ways"

foxish replied:
Done

jiangxb1987 commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,496 @@
+
+# Configuration
+
+See the [configuration page](configuration.html) for information on Spark configurations.  The following configuration is
```
" The following configuration is" -> " The following configurations are"

foxish replied:
Done

jiangxb1987 reviewed (commented):
Mostly LGTM, only some nits.

foxish commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,496 @@
+* This will become a table of contents (this text will be scraped).
+{:toc}
+
+Spark can run on clusters managed by [Kubernetes](https://kubernetes.io). This features makes use of the new experimental native
```
Done

foxish commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,496 @@
+spark-submit can be directly used to submit a Spark application to a Kubernetes cluster. The mechanism by which spark-submit happens is as follows:
+
+* Spark creates a spark driver running within a [Kubernetes pod](https://kubernetes.io/docs/concepts/workloads/pods/pod/).
+* The driver creates executors which are also Kubernetes pods and connects to them, and executes application code.
```
Done

foxish commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,496 @@
+master string with `k8s://` will cause the Spark application to launch on the Kubernetes cluster, with the API server
+being contacted at `api_server_url`. If no HTTP protocol is specified in the URL, it defaults to `https`. For example,
+setting the master to `k8s://example.com:443` is equivalent to setting it to `k8s://https://example.com:443`, but to
+connect without TLS on a different port, the master would be set to `k8s://http://example.com:8080`.
```
It should just be handled by the fabric8 client. We added this in https://github.com/fabric8io/kubernetes-client/pull/652

foxish commented on sbin/build-push-docker-images.sh:
```diff
@@ -0,0 +1,67 @@
+    push) push;;
+    *) usage;;
+  esac
+fi
```
Done

foxish commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,496 @@
+
+To launch Spark Pi in cluster mode,
+
+    bin/spark-submit \
```
Done

foxish committed:
Address comments

foxish committed:
Adding links to running-on-kubernetes.md

foxish commented:
@jiangxb1987, thanks for the round of review - I updated some links and pointers to the docs as well. PTAL.

vanzin commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+[kubectl](https://kubernetes.io/docs/user-guide/prereqs/).  If you do not already have a working Kubernetes cluster,
+you may setup a test cluster on your local machine using
+[minikube](https://kubernetes.io/docs/getting-started-guides/minikube/).
+  * We recommend using the latest releases of minikube be updated to the most recent version with the DNS addon enabled.
```
This sentence doesn't read right.

vanzin commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+  <img src="img/k8s-cluster-mode.png" title="Spark cluster components" alt="Spark cluster components" />
+</p>
+
+spark-submit can be directly used to submit a Spark application to a Kubernetes cluster. The mechanism by which spark-submit happens is as follows:
```
`<code>spark-submit</code>`

"The submission mechanism works as follows:"

foxish replied:
Done

vanzin commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+* Spark creates a spark driver running within a [Kubernetes pod](https://kubernetes.io/docs/concepts/workloads/pods/pod/).
+* The driver creates executors which are also running within Kubernetes pods and connects to them, and executes application code.
+* When the application completes, the executor pods terminate and are cleaned up, but the driver pod persists
+logs and remains in "completed" state in the Kubernetes API till it's eventually garbage collected or manually cleaned up.
```
s/till/until

foxish replied:
Done

vanzin commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+
+Kubernetes has the concept of [namespaces](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/).
+Namespaces are ways to divide cluster resources between multiple users (via resource quota). Spark on Kubernetes can
+use namespaces to launch spark applications. This is through the `--conf spark.kubernetes.namespace` argument to spark-submit.
```
Instead of mentioning it as a spark-submit argument, mention it as a configuration. Users have more than one way to set configuration in their apps.

foxish replied:
Done

vanzin commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+service account that has the right role granted. Spark on Kubernetes supports specifying a custom service account to
+be used by the driver pod through the configuration property
+`spark.kubernetes.authenticate.driver.serviceAccountName=<service account name>`. For example to make the driver pod
+to use the `spark` service account, a user simply adds the following option to the `spark-submit` command:
```
"...to make the driver pod use..."

foxish replied:
Done

vanzin commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+</tr>
+<tr>
+  <td><code>spark.kubernetes.driver.docker.image</code></td>
+  <td><code>(none)</code></td>
```
Is this a required config, since its default is unset?

foxish replied:
Yes, it's required. Same for the executor docker images. Adding a line that it's a required field.

vanzin commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+  </td>
+</tr>
+<tr>
+  <td><code>spark.kubernetes.executor.docker.image</code></td>
```
Same question.

foxish replied:
Yes, it's required.

vanzin commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+</tr>
+<tr>
+  <td><code>spark.kubernetes.allocation.batch.delay</code></td>
+  <td><code>1</code></td>
```
Hmm, isn't this a time conf? If so, specify the value with the unit and don't mention units in the description.

foxish replied:
This is an int conf at the moment - but good point about making it a time config. I think it makes sense to change that. Will draft up a PR.

foxish replied:
Fix in #20032

vanzin commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+  </td>
+</tr>
+<tr>
+  <td><code>spark.kubernetes.authenticate.submission.oauthToken</code></td>
```
If not yet supported, I recommend adding support for reading this from an env variable; command line args are visible by anyone using `ps`.

foxish replied:
That makes sense. I think the fabric8 client has a way of consuming that from env-vars, but we need to check that it works.

cc @mccheah

foxish replied:
Filing [an issue](https://issues.apache.org/jira/browse/SPARK-22842) to unblock. I think we can work around this for now - but will be important to document for the future.

vanzin commented on docs/submitting-applications.md:
```diff
@@ -155,6 +165,12 @@ The master URL passed to Spark can be in one of the following formats:
         <code>client</code> or <code>cluster</code> mode depending on the value of <code>--deploy-mode</code>.
         The cluster location will be found based on the <code>HADOOP_CONF_DIR</code> or <code>YARN_CONF_DIR</code> variable.
 </td></tr>
+<tr><td> <code>k8s://HOST:PORT</code> </td><td> Connect to a <a href="running-on-kubernetes.html"> Kubernetes </a> cluster in
```
Remove spaces around "Kubernetes".

foxish replied:
Done

vanzin commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+   <td><code>spark.kubernetes.driver.secrets.[SecretName]</code></td>
+   <td>(none)</td>
+   <td>
+     Mounts the Kubernetes secret named <code>SecretName</code> onto the path specified by the value
```
What is a "Kubernetes secret"? Do you need a link to explain it somewhere? When would you use this?

foxish replied:
Done

ueshin commented:
ok to test

SparkQA commented:
**[Test build #84850 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84850/testReport)** for PR 19946 at commit [`5f24de1`](https://github.com/apache/spark/commit/5f24de19f038ac55015f9da54cfc681b18508d24).

ueshin commented on docs/running-on-kubernetes.md:
````diff
@@ -0,0 +1,498 @@
+Kubernetes master is running at http://127.0.0.1:6443
+```
+
+In the above example, the specific Kubernetes cluster can be used with spark submit by specifying
````
`spark-submit` instead of `spark submit`?

foxish replied:
Done

ueshin commented on docs/running-on-kubernetes.md:
````diff
@@ -0,0 +1,498 @@
+The local proxy can be started by:
+
+```bash
+ kubectl proxy
````
nit: we can remove extra space at the beginning of this line.

foxish replied:
Done

ueshin commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+
+### Accessing Logs
+
+Logs can be accessed using the kubernetes API and the `kubectl` CLI. When a Spark application is running, it's possible
```
nit: `Kubernetes API` instead of `kubernetes API` (using a capital letter `K`)?

foxish replied:
done

ueshin commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+
+There may be several kinds of failures. If the Kubernetes API server rejects the request made from spark-submit, or the
+connection is refused for a different reason, the submission logic should indicate the error encountered. However, if there
+are errors during the running of the application, often, the best way to investigate may be through the kubernetes CLI.
```
nit: `Kubernetes CLI` instead of `kubernetes CLI`?

foxish replied:
Done

ueshin commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+
+Status and logs of failed executor pods can be checked in similar ways. Finally, deleting the driver pod will clean up the entire spark 
+application, includling all executors, associated service, etc. The driver pod can be thought of as the Kubernetes representation of 
+the spark application.
```
nit: `Spark application` instead of `spark application`?

foxish replied:
Done

ueshin commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+
+Kubernetes has the concept of [namespaces](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/).
+Namespaces are ways to divide cluster resources between multiple users (via resource quota). Spark on Kubernetes can
+use namespaces to launch spark applications. This is through the `--conf spark.kubernetes.namespace` argument to spark-submit.
```
ditto.

ueshin commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+  <td><code>spark.kubernetes.authenticate.driver.oauthToken</code></td>
+  <td>(none)</td>
+  <td>
+    OAuth token to use when authenticating against the against the Kubernetes API server from the driver pod when
```
`against the against` -> `against`?

foxish replied:
Done

foxish commented on docs/index.md:
```diff
@@ -112,7 +113,7 @@ options for deployment:
       [Apache Mesos](http://mesos.apache.org)
   * [YARN](running-on-yarn.html): deploy Spark on top of Hadoop NextGen (YARN)
-  * [Kubernetes (experimental)](https://github.com/apache-spark-on-k8s/spark): deploy Spark on top of Kubernetes
+  * [Kubernetes (experimental)](running-on-kubernetes.html): deploy Spark on top of Kubernetes
```
Done

SparkQA commented:
**[Test build #84850 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84850/testReport)** for PR 19946 at commit [`5f24de1`](https://github.com/apache/spark/commit/5f24de19f038ac55015f9da54cfc681b18508d24).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/84850/
Test PASSed.

foxish committed:
Addressed comments (round 2)

mridulm commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+  <td>
+    In cluster mode, whether to wait for the application to finish before exiting the launcher process.  When changed to
+    false, the launcher has a "fire-and-forget" behavior when launching the Spark job.
+  </td>
```
When false, will terminating spark submit cause spark application to also terminate ?

foxish replied:
No, it wouldn't. It would simply stop spark-submit from continuing to watch for status from the cluster. The application runs the same in either case.

mridulm reviewed (commented):
I left a few minor comments, overall looks pretty good - thanks for the PR !

SparkQA commented:
**[Test build #84861 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84861/testReport)** for PR 19946 at commit [`f618e8b`](https://github.com/apache/spark/commit/f618e8b63a3a33bae0ad95617298c538cb13adb7).

foxish commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+decisions for driver and executor pods using advanced primitives like
+[node selectors](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#nodeselector)
+and [node/pod affinities](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#affinity-and-anti-affinity)
+in a future release.
```
As of today, preemption is at random. 
[priority and preemption](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/) are in alpha as of now. As soon as they go to beta (in the Spark 2.4 timeframe), we'll add the required pieces to make it honor the rule as you said - driver is the last to go, etc.

foxish commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+{:toc}
+
+Spark can run on clusters managed by [Kubernetes](https://kubernetes.io). This feature makes use of the new experimental native
+Kubernetes scheduler that has been added to Spark.
```
Done. The other references have been removed. Couldn't find any others.

foxish commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+$ bin/spark-submit \
+    --deploy-mode cluster \
+    --class org.apache.spark.examples.SparkPi \
+    --master k8s://https://<k8s-apiserver-host>:<k8s-apiserver-port> \
```
Done

foxish commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+    --deploy-mode cluster \
+    --class org.apache.spark.examples.SparkPi \
+    --master k8s://https://<k8s-apiserver-host>:<k8s-apiserver-port> \
+    --conf spark.kubernetes.namespace=default \
```
Removed

foxish committed:
review comments

foxish commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+    --conf spark.app.name=spark-pi \
+    --conf spark.kubernetes.driver.docker.image=<driver-image> \
+    --conf spark.kubernetes.executor.docker.image=<executor-image> \
+    local:///opt/spark/examples/jars/spark-examples_2.11-2.3.0.jar
```
Fair enough, I just wanted to show that it had to be a path. The protocol `local:///` is necessary to point at container-local files, as opposed to local file paths on the submitting user's machine which will both be handled separately. Changed it to `local:///path/to/examples.jar`. Is that better?

mridulm replied:
Just to clarify - local:///my/path/jar implies local to the docker container ? and /my/path/jar points to submitters machine ?

SparkQA commented:
**[Test build #84864 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84864/testReport)** for PR 19946 at commit [`8c656c5`](https://github.com/apache/spark/commit/8c656c56a44e3f811cb376dd3ff013643a5f7206).

foxish commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+    --class org.apache.spark.examples.SparkPi \
+    --master k8s://https://<k8s-apiserver-host>:<k8s-apiserver-port> \
+    --conf spark.kubernetes.namespace=default \
+    --conf spark.executor.instances=5 \
```
That is specified currently as a YARN-only option in `spark-submit`. Does it make sense to move it out to suit both K8s and YARN?

mridulm replied:
You are right, I did not realize mesos did not honour it and it was yarn only option !
It might make sense for k8s also to support it too (since it does not right now, my proposal for doc change would be incorrect) .. what do you think @foxish  ?

foxish committed:
Update to --name

foxish commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+    --master k8s://https://<k8s-apiserver-host>:<k8s-apiserver-port> \
+    --conf spark.kubernetes.namespace=default \
+    --conf spark.executor.instances=5 \
+    --conf spark.app.name=spark-pi \
```
Done

SparkQA commented:
**[Test build #84865 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84865/testReport)** for PR 19946 at commit [`109069f`](https://github.com/apache/spark/commit/109069f80be8c3ecd379643300173295ea35bb59).

mridulm commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,498 @@
+* When the application completes, the executor pods terminate and are cleaned up, but the driver pod persists
+logs and remains in "completed" state in the Kubernetes API till it's eventually garbage collected or manually cleaned up.
+
+Note that in the completed state, the driver pod does *not* use any computational or memory resources.
```
Some edit probably buried this query ... would be great if you can opine @foxish, thanks.

SparkQA commented:
**[Test build #84865 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84865/testReport)** for PR 19946 at commit [`109069f`](https://github.com/apache/spark/commit/109069f80be8c3ecd379643300173295ea35bb59).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/84865/
Test FAILed.

SparkQA commented:
**[Test build #84861 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84861/testReport)** for PR 19946 at commit [`f618e8b`](https://github.com/apache/spark/commit/f618e8b63a3a33bae0ad95617298c538cb13adb7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/84861/
Test PASSed.

SparkQA commented:
**[Test build #84864 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84864/testReport)** for PR 19946 at commit [`8c656c5`](https://github.com/apache/spark/commit/8c656c56a44e3f811cb376dd3ff013643a5f7206).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/84864/
Test PASSed.

rxin commented on docs/building-spark.md:
```diff
@@ -49,7 +49,7 @@ To create a Spark distribution like those distributed by the
 with Maven profile settings and so on like the direct Maven build. Example:
 
-    ./dev/make-distribution.sh --name custom-spark --pip --r --tgz -Psparkr -Phadoop-2.7 -Phive -Phive-thriftserver -Pmesos -Pyarn
+    ./dev/make-distribution.sh --name custom-spark --pip --r --tgz -Psparkr -Phadoop-2.7 -Phive -Phive-thriftserver -Pmesos -Pyarn -Pkubernetes
```
should we use k8s? I kept bringing this up and that's because I can never spell Kubernetes properly.

foxish replied:
It's `k8s://` in the URL scheme we use and the package names are also `k8s` - so, users should never have to type the name.
This is one of the last few holdouts. I'd say that it's consistent here, with the use of other cluster manager names in full in their maven projects. I can change it here if you feel strongly about it.

mridulm replied:
I actually second @rxin, I do get the spelling wrong at times :-)

foxish replied:
Are you both also referring to the config options etc, which are still `spark.kubernetes.*`, or just the maven build target? If it's everything, it would be a fairly large change.

foxish replied:
Ping @mridulm, @rxin - can you please confirm the scope of the renaming you were referring to here? Is it just the maven target? Changing all the config options etc would be a considerably large change at this point. Also, a point that was brought up today was - while k8s is common shorthand, it's not universal as is the full name.

foxish replied:
I've filed https://issues.apache.org/jira/browse/SPARK-22853 to discuss this and unblock this PR. We should be able to reach consensus by release time. :)

rxin replied:
Yea I don't think you need to block this pr with this.

AmplabJenkins commented:
Can one of the admins verify this patch?

vanzin commented:
Seems like the PR builders lost their state?

vanzin commented:
ok to test

SparkQA commented:
**[Test build #84929 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84929/testReport)** for PR 19946 at commit [`109069f`](https://github.com/apache/spark/commit/109069f80be8c3ecd379643300173295ea35bb59).

vanzin commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,502 @@
+## Docker Images
+
+Kubernetes requires users to supply images that can be deployed into containers within pods. The images are built to
+be run in a container runtime environment that Kubernetes supports. Docker is a container runtime environment that is
```
> a container runtime environment that Kubernetes supports

Does the current code support these other container runtimes, or just docker? From my memory (and the propery names) it seems that only docker is supported?

erikerlandson replied:
the images should be runnable by other container runtimes (cri-o, rkt, etc), although I'm not aware of anybody testing that.

foxish replied:
@vanzin, we don't make any runtime level assumptions in Spark code. The K8s abstraction layer [CRI](http://blog.kubernetes.io/2016/12/container-runtime-interface-cri-in-kubernetes.html) should in theory allow using a differnet runtime.

vanzin replied:
My comment is more about the properties being called "docker" and whether that means only docker images are supported. If you can use any image supported by the k8s cluster, than pehaps the properties should be renamed.

foxish replied:
I see your point - Although there is flexibility in theory, as of now, it's safe to assume that *most* people are running docker containers when using k8s - making the name docker much more intuitive. If the other runtimes do see traction in future (and we do some testing around them), we can rename to `container.image` instead of `docker.image`. As of now, I can make the documentation clearer that spark on k8s only supports docker images. Sound like a reasonable thing to do here?

jiangxb1987 replied:
Just my two cents, if we could foresee that we may add support to other containers, then we should consider using `container.image` instead of `docker.image` at the beginning, because rename a config is also considered a kind of behavior change, that commonly takes more effort to process. If it is not that case (we are satisfied to be only supporting docker containers), then it would be fine to just keep the `docker.image` phrase.

ueshin replied:
I wonder if k8s can tell the container runtime from the image name? If it can, we can use `container.image` here, but otherwise, I guess we need another config like `xxx.container` to tell the container runtime and `xxx.${container}.image` to specify the image, e.g. `xxx.container=docker` and `xxx.docker.image=something`.

vanzin replied:
+1 to @ueshin's comment; if the same property supports more than docker images, it should be renamed to use a more generic name.

foxish replied:
Makes sense. I think `container.image` should be fine then - will send a PR changing that. 
I don't think there are plans to support choosing the runtime on the fly, or letting the CRI abstraction leak into the application layer.

erikerlandson replied:
I'd expect the container run-time to be transparent to the particular image, and very definitely transparent to spark.  Using `container.image` makes sense from either standpoint.

foxish replied:
https://github.com/apache/spark/pull/19995 PTAL

vanzin commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,502 @@
+    ./sbin/build-push-docker-images.sh -r <repo> -t my-tag build
+    ./sbin/build-push-docker-images.sh -r <repo> -t my-tag push
+
+Docker files are under the `dockerfiles/` and can be customized further before
```
"the `dockerfiles/` directory in the Spark distribution archive"?

(Assuming that's true. Need to clarify where to find `dockerfiles/`.)

foxish replied:
It is not yet true - it's a TODO item on this PR. Will clarify after we get that change in.

foxish replied:
This has been addressed and the script has been fixed now.

vanzin commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,502 @@
+
+## Dependency Management
+
+If your application's dependencies are all hosted in remote locations like HDFS or http servers, they may be referred to
```
"HTTP servers"

foxish replied:
Done

vanzin commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,502 @@
+  <td>
+    Docker image to use for the executors. Specify this using the standard
+    <a href="https://docs.docker.com/engine/reference/commandline/tag/">Docker tag</a> format.
+    This configuration is required and must be provided by the user.
```
Would it make sense to have this default to the same image used for the driver? Or are they fundamentally different in some way?

erikerlandson replied:
They use different commands to run driver vs executor, and also have some differing environment expectations, so defaulting them to the same would not work

vanzin commented on docs/running-on-yarn.md:
```diff
@@ -18,7 +18,8 @@ Spark application's configuration (driver, executors, and the AM when running in
 There are two deploy modes that can be used to launch Spark applications on YARN. In `cluster` mode, the Spark driver runs inside an application master process which is managed by YARN on the cluster, and the client can go away after initiating the application. In `client` mode, the driver runs in the client process, and the application master is only used for requesting resources from YARN.
 
-Unlike [Spark standalone](spark-standalone.html) and [Mesos](running-on-mesos.html) modes, in which the master's address is specified in the `--master` parameter, in YARN mode the ResourceManager's address is picked up from the Hadoop configuration. Thus, the `--master` parameter is `yarn`.
+Unlike [Spark standalone](spark-standalone.html), [Mesos](running-on-mesos.html) and [Kubernetes](running-on-kubernetes.html) modes,
```
Should just say "other cluster managers supported by Spark" at this point.

foxish replied:
Done

foxish committed:
Review comments

SparkQA commented:
**[Test build #84934 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84934/testReport)** for PR 19946 at commit [`873f04d`](https://github.com/apache/spark/commit/873f04dc3c1a4c48cbeac5766aec7c23c810c0cb).

SparkQA commented:
**[Test build #84929 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84929/testReport)** for PR 19946 at commit [`109069f`](https://github.com/apache/spark/commit/109069f80be8c3ecd379643300173295ea35bb59).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/84929/
Test PASSed.

SparkQA commented:
**[Test build #84934 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/84934/testReport)** for PR 19946 at commit [`873f04d`](https://github.com/apache/spark/commit/873f04dc3c1a4c48cbeac5766aec7c23c810c0cb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/84934/
Test PASSed.

felixcheung commented:
pending updates from #19995 and #20007

foxish committed:
Update the image building script

foxish committed:
Add path to dockerfiles

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85166/
Test FAILed.

SparkQA commented:
**[Test build #85167 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85167/testReport)** for PR 19946 at commit [`74ac5c9`](https://github.com/apache/spark/commit/74ac5c9e5b495d0133e8e1378867a43f2bc1ff4a).

foxish committed:
Test and update various options

SparkQA commented:
**[Test build #85178 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85178/testReport)** for PR 19946 at commit [`d235847`](https://github.com/apache/spark/commit/d2358470e86ab44522371b8fd733a97527d95ec5).

foxish committed:
Add <code>..</code> blocks

SparkQA commented:
**[Test build #85179 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85179/testReport)** for PR 19946 at commit [`702162b`](https://github.com/apache/spark/commit/702162b4ca9eab83adb0b362d5b4d9479b6b3d0a).

SparkQA commented:
**[Test build #85167 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85167/testReport)** for PR 19946 at commit [`74ac5c9`](https://github.com/apache/spark/commit/74ac5c9e5b495d0133e8e1378867a43f2bc1ff4a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85167/
Test PASSed.

SparkQA commented:
**[Test build #85178 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85178/testReport)** for PR 19946 at commit [`d235847`](https://github.com/apache/spark/commit/d2358470e86ab44522371b8fd733a97527d95ec5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85178/
Test PASSed.

SparkQA commented:
**[Test build #85179 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85179/testReport)** for PR 19946 at commit [`702162b`](https://github.com/apache/spark/commit/702162b4ca9eab83adb0b362d5b4d9479b6b3d0a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85179/
Test PASSed.

liyinan926 commented on docs/running-on-kubernetes.md:
```diff
@@ -0,0 +1,573 @@
+$ bin/spark-submit \
+    --master k8s://https://<k8s-apiserver-host>:<k8s-apiserver-port> \
+    --deploy-mode cluster \
+    --name spark-pi \
```
We need to explicitly call out that app names in k8s mode cannot have spaces and special characters.

foxish replied:
Good point, will update with caveat.

foxish replied:
Done

vanzin commented:
BTW can I ask you guys to start using "K8S" in the PR titles? These are not really scheduler changes.

foxish commented:
Makes sense, will be sure to categorize appropriately moving forward.
Thanks!

On Dec 20, 2017 4:16 PM, "Marcelo Vanzin" <notifications@github.com> wrote:

BTW can I ask you guys to start using "K8S" in the PR titles? These are not
really scheduler changes.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/apache/spark/pull/19946#issuecomment-353220449>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AA3U53sCa7IHsWMCjwYxlIWBCmkDrqUKks5tCaNYgaJpZM4Q-PQe>
.

foxish committed:
Update instructions for spark.app.name, modify config type

SparkQA commented:
**[Test build #85222 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85222/testReport)** for PR 19946 at commit [`8726154`](https://github.com/apache/spark/commit/87261549dd41c6a8c97a5d2fc33119c0296265e4).

foxish commented:
Addressed comments. This PR should be ready to go, with one separate issue of renaming to be discussed inhttps://issues.apache.org/jira/browse/SPARK-22853. 

PTAL @vanzin @mridulm @ueshin @jiangxb1987 @felixcheung @rxin 
This covers 90% of our documentation for Spark 2.3 (pending the documentation needed for https://github.com/apache/spark/pull/19954).

SparkQA commented:
**[Test build #85222 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85222/testReport)** for PR 19946 at commit [`8726154`](https://github.com/apache/spark/commit/87261549dd41c6a8c97a5d2fc33119c0296265e4).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85222/
Test FAILed.

jiangxb1987 commented on docs/running-on-yarn.md:
```diff
@@ -18,7 +18,8 @@ Spark application's configuration (driver, executors, and the AM when running in
 
-Unlike [Spark standalone](spark-standalone.html) and [Mesos](running-on-mesos.html) modes, in which the master's address is specified in the `--master` parameter, in YARN mode the ResourceManager's address is picked up from the Hadoop configuration. Thus, the `--master` parameter is `yarn`.
+Unlike other cluster managers supported by Spark
+in which the master's address is specified in the `--master` parameter, in YARN mode the ResourceManager's address is picked up from the Hadoop configuration. Thus, the `--master` parameter is `yarn`.
```
nit: why start a new line here?

foxish replied:
Done

foxish committed:
Better line breaks

SparkQA commented:
**[Test build #85278 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85278/testReport)** for PR 19946 at commit [`374ddc8`](https://github.com/apache/spark/commit/374ddc850e06f9a5d5d83ddc32b3e28e5fbf1007).

SparkQA commented:
**[Test build #85278 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85278/testReport)** for PR 19946 at commit [`374ddc8`](https://github.com/apache/spark/commit/374ddc850e06f9a5d5d83ddc32b3e28e5fbf1007).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85278/
Test PASSed.

foxish commented:
Ping - any new comments?

jiangxb1987 commented:
LGTM

rxin commented:
Merging in master.
